minecode/visitors/debian.py

```diff
@@ -298,3 +316,380 @@ class DebianReleaseVisitor(HttpVisitor):
+        'other_license_detections',
+        'extracted_license_statement'
+    ]
+
+    for field in fields_to_update:
+        value = getattr(package_from, field)
+        if value and replace:
+            setattr(package_to, field, value)
+
+
+def map_debian_metadata_binary_and_source(package_url, source_package_url):
+    """
+    Get metadata for the binary and source release of the Debain package
+    `package_url` and save it to the PackageDB.
+
+    Return an error string for errors that occur, or empty string if there is no error.
+    """
+    error = ''
+
+    if "repository_url" in package_url.qualifiers:
+        base_url = package_url.qualifiers["repository_url"]
+    elif package_url.namespace == 'ubuntu':
+        base_url = UBUNTU_BASE_URL
+    else:
+        base_url = DEBIAN_BASE_URL
+    
+    if "api_data_url" in package_url.qualifiers:
+        metadata_base_url = package_url.qualifiers["api_data_url"]
+    elif package_url.namespace == 'ubuntu':
+        metadata_base_url = UBUNTU_METADATA_URL
+    else:
+        metadata_base_url = DEBIAN_METADATA_URL
+
+    debian_package = DebianPackage(
+        package_url=package_url,
+        source_package_url=source_package_url,
+        archive_base_url=base_url,
+        metadata_base_url=metadata_base_url,
+    )
+
+    binary_package, emsg = map_debian_package(
+        debian_package,
+        PackageContentType.BINARY,
+    )
+    if emsg:
+        error += emsg
+
+    package_url.qualifiers['classifier'] = 'sources'
+    source_package, emsg = map_debian_package(
+        debian_package,
+        PackageContentType.SOURCE_ARCHIVE,
+    )
+    if emsg:
+        error += emsg
+
+    if binary_package and source_package:
+        make_relationship(
+            from_package=binary_package,
+            to_package=source_package,
+            relationship=PackageRelation.Relationship.SOURCE_PACKAGE,
+        )
+
+    return error
+
+
+@attr.s
+class DebianPackage:
+
+    archive_base_url = attr.ib(type=str)
+    metadata_base_url = attr.ib(type=str)
+    package_url = attr.ib(type=str)
+    source_package_url = attr.ib(type=str)
+    metadata_directory_url = attr.ib(type=str, default=None)
+    archive_directory_url = attr.ib(type=str, default=None)
+
+    def __attrs_post_init__(self, *args, **kwargs):
+        self.set_debian_directories()
+
+    @property
+    def package_archive_version(self):
+        """
+        """
+        debvers = DebVersion.from_string(self.package_url.version)
+        if debvers.revision != "0":
+            purl_version = f"{debvers.upstream}-{debvers.revision}"
+        else:
+            purl_version = debvers.upstream
+        return purl_version
+
+    @property
+    def binary_archive_url(self):
+        """
+        """
+        purl_version = self.package_archive_version
+        arch = self.package_url.qualifiers.get("architecture")
+        if arch:
+            archive_name =f"{self.package_url.name}_{purl_version}_{arch}.deb"
```

Comment:
This is not always correct. There are some packages in Ubuntu that have the extension `.udeb`, instead of `.deb`. An example is libzstd, which includes both `.deb` and `.udeb` versions.

For an explanation of `.udeb`: https://wiki.debian.org/udeb

@pombredanne what do you think?

Reply:
Thanks for this @armijnhemel , please let us know if there are other things we are missing here.
From the link:
> An udeb is a stripped down [deb](https://wiki.debian.org/deb) file for use by the [DebianInstaller](https://wiki.debian.org/DebianInstaller) . It removes the naughty bits (the documentation, as well as some checksums, etc.) to save space.

1. Are there cases where there is only a `.udeb` file and no `.deb` files for a package?
2. If both `.udeb` and `.deb` files are present, is it not better to return the `.deb` file as the main binary archive?
3. Currently I was only using packages detected in the latest debian docker images at: `docker://debian:bullseye` and `docker://ubuntu:devel` as test cases for the purls (note that sometimes older versions are yanked from debian, need to support these from the archives too), anything else that would be nice to test?

Reply:
> Thanks for this @armijnhemel , please let us know if there are other things we are missing here. From the link:
> 
> > An udeb is a stripped down [deb](https://wiki.debian.org/deb) file for use by the [DebianInstaller](https://wiki.debian.org/DebianInstaller) . It removes the naughty bits (the documentation, as well as some checksums, etc.) to save space.
> 
>     1. Are there cases where there is only a `.udeb` file and no `.deb` files for a package?

Yes, there are. A few examples I could find (both from Ubuntu):

https://ftp.nluug.nl/pub/os/Linux/distr/ubuntu/pool/main/d/debian-installer-utils/
https://ftp.nluug.nl/pub/os/Linux/distr/ubuntu/pool/main/u/udpkg/

> 
>     2. If both `.udeb` and `.deb` files are present, is it not better to return the `.deb` file as the main binary archive?

I am not sure. Both are an instance derived from the same source code, so both would likely be valid matches. As these packages are strongly tied to the installer and unlikely present on a running system like a container, I could imagine that the `.deb` is probably a better match almost always, but perhaps not always. Also, the packages seem to have different names according to the Purl syntax. Let's look at:

https://ftp.nluug.nl/pub/os/Linux/distr/ubuntu/pool/main/b/busybox/

There are a few packages:

* `busybox`
* `busybox-static`
* `busybox-udeb`

These would all get different purls, but would have been derived from the same source code (also see #308 which is somehwat related to this).

@pombredanne : thoughts?



> 
>     3. Currently I was only using packages detected in the latest debian docker images at: `docker://debian:bullseye` and `docker://ubuntu:devel` as test cases for the purls (note that sometimes older versions are yanked from debian, need to support these from the archives too), anything else that would be nice to test?

I am not sure if all the old versions can even be found in the archives. I cannot think of anything else to test.

Reply:
AFAIK, you cannot be both a deb and a udeb at once. There is a field in the control file that drives this: `Package-Type`

A udeb package and a deb can be built possibly built from the same source like for busybox, but are different packages:
- https://ftp.nluug.nl/pub/os/Linux/distr/ubuntu/pool/main/b/busybox/busybox_1.21.0-1ubuntu1.4.dsc is for the source
- `busybox-udeb_1.21.0-1ubuntu1.4_amd64.udeb` is the one built as udeb
- the control file in https://ftp.nluug.nl/pub/os/Linux/distr/ubuntu/pool/main/b/busybox/busybox_1.21.0-1ubuntu1.4.debian.tar.gz has :+1: 
```
Package: busybox-udeb
Package-Type: udeb
```
We could use a "package_type" as a qualifier. We should track the implementation and support for udeb in a separate issue IMHO.

Reply:
> I am not sure if all the old versions can even be found in the archives. I cannot think of anything else to test.

They are in https://snapshot.debian.org/ FWIW